Basic implementation of Zwave Rollershutters

turbokongen opened the pull request:
**Description:**
Basic Implementation for Zwave Rollershutter devices.

This also blocks tested rollershutter device from switch and light.
This is only tested with Fibaro FGRM-222 rollershutter device, and should be considered a basic implementation. For the future, other devices may need to be added for blocking.
Thank you @nunofgs for providing the information to make this possible, testing and debug. :+1: 
Thank you @wokar for providing code for better filtering. This will make future filtration easier. :+1: 
**Related issue (if applicable):** fixes #
#2053

**Checklist:**

If code communicates with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

turbokongen committed:
Basic implementation of Zwave Rollershutters

balloob commented on homeassistant/components/zwave.py:
```diff
@@ -92,6 +93,10 @@
      TYPE_BOOL,
      GENRE_USER),
+    ('rollershutter',
+     [COMMAND_CLASS_SWITCH_MULTILEVEL],
```
Doesn't this mean that all multilevel switches will be forwarded to the rollershutter component ?

turbokongen replied:
No, it get filtered by the next two arguments. The index type and genre.

turbokongen replied:
Type_whatever and genre_whatever would have done that.

turbokongen replied:
Ofcourse untested :pray: , neither of us have that combination.
Anyone that do have a dimmer and a rollershutter device that can test?

turbokongen replied:
Are there any other ways to channel this correctly. I see there will be problems with identifying devices correctly to the correct component. I understood the as this:
If command class is switch multilevel AND contains a value with button AND has genre with user, it will be detected as rollershutter. 
Or have I misunderstood this? Open for suggestions :)

wokar replied:
I started to implement rollershutter support myself a few days ago - but it you were faster ;-) My approach was to add the "specific device class" as another property to identify devices. For the Fibaro rollershutter it has to be MOTOR_CONTROL_CLASS_B. This way dimmers will always be detected as lights, but on the other hand things like garage door openers might also be detected as rollershutters. So its still not perfect, but to detect every motor controller as a rollershutter is still much better than to detect it as a light IMHO. Just take a look at my forks zwave-rollershutter branch...

turbokongen replied:
I was looking at the specific command class last night. Excellent work😊.It looks like a good idea for the future too. Feel free to make a pull request against my branch, as I won't have time till Saturday to change this. With the code you have at the Base of zwave.py it should be easier to filter components.👍

turbokongen replied:
I got time to add your changes. Thank you @wokar :+1:

turbokongen replied:
@wokar  Is there a list somewhere that we can use to find what specific command class is mapped to which number?

wokar replied:
@turbokongen  For the zwave constants I use the openhab project as a reference - the  [ZWaveCommandClass.java](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveCommandClass.java) contains the mapping you want.

turbokongen replied:
Yes, that list I am aware of, but the specific class is what I am interessted in. In this case MOTOR_CONTROL_CLASS_B is 6. Is there a list that those mappings are available?

wokar replied:
Ah yes - here the [openhab](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/ZWaveDeviceClass.java) class - or the [openzwave](https://github.com/wokar/open-zwave/blob/master/config/device_classes.xml) mapping which better shows the relation of generic, specific and required command classes.

turbokongen committed:
Better filtering, by @wokar

turbokongen committed:
Fix typo

balloob commented on homeassistant/components/light/zwave.py:
```diff
@@ -20,6 +32,15 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     node = zwave.NETWORK.nodes[discovery_info[zwave.ATTR_NODE_ID]]
     value = node.values[discovery_info[zwave.ATTR_VALUE_ID]]
+    # Make sure that we have values for the key before converting to int
+    if (value.node.manufacturer_id.strip() and
```
We should really create a helper method in the `zwave` component that you pass in your `value` and `WORKAROUNDS` constant and it will return `True` or `False`

turbokongen replied:
Uhm.... Sure. Unsure how to go about that. Because this will eventually be present for virtually all components zwave is loaded on. When things evolve it will get harder to send devices to the correct component. So some routine to recognize the correct is needed, This needs some investegation to get correct.

balloob replied:
In the Z-Wave module add

```python3
workaround = get_workaround(value, WORKAROUNDS)

if workaround == WORKAROUND_IGNORE:
    return
elif …
```

turbokongen replied:
I'm about to commit a different approach, basicly I have benn looking at all devices in the pepper1 database looking at the multiswitch command class, and came up with a solution to filter on Generic device class and specific device class. That combination looks to bee the easiest way to filter stuff correctly. At command Class binary switch it will be impossible to block components, because more or less all devices has that command class whatever extra capabilities it has. So switch needs to be hidden with the costumize section of the config.

balloob commented on homeassistant/components/light/zwave.py:
```diff
@@ -20,6 +32,15 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
+                               int(value.node.product_id, 16))
+        if specific_sensor_key in DEVICE_MAPPINGS:
+            if DEVICE_MAPPINGS[specific_sensor_key] == WORKAROUND_IGNORE:
+                _LOGGER.debug("Fibaro FGRM-222 Rollershutter, ignoring")
```
This debug statement is correct now, but not when we add more rollershutters.

balloob commented on homeassistant/components/rollershutter/zwave.py:
```diff
@@ -0,0 +1,95 @@
+            _LOGGER.debug("Value changed on network %s", value)
+
+    @property
+    def name(self):
```
Why overwrite the `name` property from `zwave.ZWaveDeviceEntity`?

turbokongen replied:
Just to get rid of the index name of the recognized value. Int this case, the rollershutterdevice would be named: `Fibaro system FGRM222 Roller Shutter Controller Open <node number>`
With this mod, we get: `Fibaro system FGRM222 Roller Shutter Controller`
If this is unwanted, I will remove it.

balloob replied:
Didn't we add the node value to make sure all names are unique ?

turbokongen replied:
True. I will remove it :)

balloob commented on homeassistant/components/rollershutter/zwave.py:
```diff
@@ -0,0 +1,95 @@
+            self._value.node.manufacturer_name, self._value.node.product_name)
+
+    @property
+    def should_poll(self):
```
No need to overwrite, already defined in `zwave.ZWaveDeviceEntity`

balloob commented on homeassistant/components/rollershutter/zwave.py:
```diff
@@ -0,0 +1,95 @@
+        """Move the roller shutter up."""
+        for value in self._node.get_values(
+                class_id=COMMAND_CLASS_SWITCH_MULTILEVEL).values():
+            if value.command_class == 38 and value.index == 0:
```
You probably want to break out of the for loop after you found the correct match?

balloob commented on homeassistant/components/switch/zwave.py:
```diff
@@ -18,6 +31,15 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
 
     node = zwave.NETWORK.nodes[discovery_info[zwave.ATTR_NODE_ID]]
     value = node.values[discovery_info[zwave.ATTR_VALUE_ID]]
+    # Make sure that we have values for the key before converting to int
```
That helper method I mentioned in a previous comment, that should happen.

turbokongen committed:
Remove polling from component, and loop fix

turbokongen committed:
linter fix

turbokongen commented:
Soooo.... we need to fix test coverage for other things too now to pass? 😄

turbokongen committed:
Filter to channel devices to correct component

turbokongen committed:
Remove overwriting of parent node name

balloob commented:
Looks good! :dolphin:

wokar commented:
@turbokongen my raspberry crashed on the weekend so I could not test your solution with my Fibaro rollershutter. Now I got the raspberry up and running again and my zwave devices are working fine, but the rollershutter seems not to work at all with the current code :( 
Before I dive into the problem - have you successfully tested your solution with a real device? If yes - then this must be a problem with my raspberry...

turbokongen commented:
I don,t have the device myself, but I had @nunofgs who does have it, help me with the testing.

turbokongen commented:
I have discovered a flaw in the discovery code, can you try with my latest:
https://github.com/turbokongen/home-assistant/tree/zwave_garagedoor

wokar commented:
Last night I figured out that the if-statement:
```python
            if specific_device_class is not None and \
               specific_device_class != node.specific:
                continue
```
is the problem. Since the specific_device_class is now a list I changed it to:
```python
            if specific_device_class is not None and \
               node.specific not in specific_device_class:
                continue
```
After that change the rollershutter showed up, but I could not control the device from the UI. I will continue testing and open a PR with my changes.

turbokongen commented:
Just download the changes I made now :)

turbokongen commented:
MAde a mistake in the code try latest.

turbokongen commented:
Ugh... Not working.

turbokongen commented:
Please test latest. https://github.com/turbokongen/home-assistant/tree/zwave_garagedoor

wokar commented:
Did a bit of cleanup of the rollershutter code, take a look at https://github.com/wokar/home-assistant/tree/zwave_rollershutter Rollershutter works fine now, execpt for the stop button. It did work the first time, but now ot is ignored or it changes direction.

I also tested your component discovery and it I had to add GENERIC_COMMAND_CLASS_BINARY_SENSOR to the 'sensor' platform to get my fibaro motion sensor working. I also changed the  *_WHATEVER entries to be not in a list, otherwise the `specific_device_class is not None` code will not work.

turbokongen commented:
Was too tired yesterday to spot my mistakes. Please test latest :)

turbokongen commented:
Please try the latest at:
https://github.com/turbokongen/home-assistant/tree/zwave_garagedoor

Hopefully all will be good now :) My zwave network atleast looks like it did before.
There have been done some additional changes to filter components,

wokar commented:
Perfect! All my devices show up and work as expected.

turbokongen commented:
:+1

nunofgs commented:
@turbokongen Hi again! I just tested the latest changes in your _zwave_garagedoor_ branch, and almost everything works as expected, except the `move_up` and `move_down` commands, which are switched!

I changed the following and everything is working perfectly:

```diff
diff --git i/homeassistant/components/rollershutter/zwave.py w/homeassistant/com
index 50ef2ed..60fa1fe 100644
--- i/homeassistant/components/rollershutter/zwave.py
+++ w/homeassistant/components/rollershutter/zwave.py
@@ -61,11 +61,11 @@ class ZwaveRollershutter(zwave.ZWaveDeviceEntity, Rollershut

     def move_up(self, **kwargs):
         """Move the roller shutter up."""
-        self._node.set_dimmer(self._value.value_id, 0)
+        self._node.set_dimmer(self._value.value_id, 100)

     def move_down(self, **kwargs):
         """Move the roller shutter down."""
-        self._node.set_dimmer(self._value.value_id, 100)
+        self._node.set_dimmer(self._value.value_id, 0)

     def stop(self, **kwargs):
```

turbokongen commented:
I fixed this issue last night in a different branch. It is already merged,
so test with Ha's dev branch. All should be good.
25. jun. 2016 14:36 skrev "Nuno Sousa" <notifications@github.com>:

> @turbokongen <https://github.com/turbokongen> Hi again! I just tested the
> latest changes in your *zwave_garagedoor* branch, and almost everything
> works as expected, except the move_up and move_down commands, which are
> switched!
>
> I changed the following and everything is working perfectly:
>
> diff --git i/homeassistant/components/rollershutter/zwave.py w/homeassistant/com
> index 50ef2ed..60fa1fe 100644--- i/homeassistant/components/rollershutter/zwave.py+++ w/homeassistant/components/rollershutter/zwave.py@@ -61,11 +61,11 @@ class ZwaveRollershutter(zwave.ZWaveDeviceEntity, Rollershut
>
>      def move_up(self, **kwargs):
>          """Move the roller shutter up."""-        self._node.set_dimmer(self._value.value_id, 0)+        self._node.set_dimmer(self._value.value_id, 100)
>
>      def move_down(self, **kwargs):
>          """Move the roller shutter down."""-        self._node.set_dimmer(self._value.value_id, 100)+        self._node.set_dimmer(self._value.value_id, 0)
>
>      def stop(self, **kwargs):
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/2313#issuecomment-228538053>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AP2koLIme1CdbFdxTM1M4uY6TQdM7Cmqks5qPSCIgaJpZM4I3WFM>
> .
>

turbokongen commented:
Oooooo!!! This is rollershutter. Wil issue a fix right away!

jbaudoux commented:
Replaced by cover component.
See https://github.com/home-assistant/home-assistant/pull/4037
